### DISCUSSION TITLE
Correct way of doing token exchange from SPA to third-in-line-API

### DISCUSSION THREAD

**Discussion by ChrisBrandhorst** [keycloak/keycloak#17440] [Q&A]
So we have the following setup:

- `SPA` is a public client, user logs in through Keycloak
- `SPA` calls backend `API1` for collecting data
- `API1` uses service `API2` on our platform for a specific functionality
- `API2` needs to use third-party `API3`
- `API1`, `API2` and `API3` are all confidential clients on Keycloak. Only `API1` is accessible from the `SPA`

We have the wish to use tokens related to the user through the entire chain. Our initial idea was as follows:

1. `SPA` calls endpoint on `API1` using access token (token `T1`) with `aud: SPA, azp: SPA`
2. `API1` exchanges token `T1` using its client credentials. Resulting token `T2` has `aud: API2, azp: API1`
3. `API1` uses token `T2` in the request to `API2`
4. `API2` exchanges token `T2` using its client credentials. Resulting token `T3` should have `aud: API3, azp: API2`
5. `API2` uses token `T3` in the request to `API3`

However, step 4 fails with Keycloak throwing an `Invalid token` error stating:
`Client session for client 'API1' not present in user session`

**First question**
I have trouble understanding that Keycloak checks if there is a session for the `azp` of the token that is being exchanged, since this (per my knowledge of the spec) merely refers to the party who _receives_ the token from Keycloak, not the party who should _use_ the token. To me, checking if there is a session for the `aud` of the session would be more logical.

Related: in the code, Keycloak referes to the `azp` by the function `getIssuedFor` (in my book, "for" being the audience), instead of the OIDC spec, which mentions "issued **to**" (in my book, the receiving party).

Maybe someone can heighten our understanding of this? Where is our thinking wrong or are we looking at an inconsistency in Keycloak?

**Second question**
What **is** the correct way to use the token of the `SPA` to get a token which can be given to `API3` in this case? For now, we have settled for starting with `SPA` exchanging token `T1` (using `T1` as Bearer) for a token with `aud: API1, azp: SPA`.

Is this the way to go? Any alternatives? Thanks for any insights.

**Comment by J0F3** [comment]
@ChrisBrandhorst we trying achieve the same setup and we have exactly the same problem. Did you find a solution at sometime or stillt the same state today?
Thx!

**Reply by ChrisBrandhorst** [reply]
@J0F3 unfortunately not. I also moved away from this project. Still interested in the correct way, though :)
Do you feel this is a bug? We might up it to an issue instead of a question then...

**Reply by J0F3** [reply]
Ok, I will keep you posted :-)

> Do you feel this is a bug? We might up it to an issue instead of a question then...
I am not sure (yet). It feels like yes. But I have also to less experience in the topic to say it for sure...

**Reply by flyinfish** [reply]
Hi @ChrisBrandhorst thanks for discussion is opened issue
https://github.com/keycloak/keycloak/issues/30614
